compiler/rustc_mir_transform/src/valgrind_client_request.rs

```diff
@@ -0,0 +1,112 @@
+use crate::MirPass;
+
+use rustc_index::vec::Idx;
+use rustc_middle::mir::patch::MirPatch;
+use rustc_middle::mir::{
+    AggregateKind, BasicBlockData, Body, Constant, ConstantKind, Location, NonDivergingIntrinsic,
+    Operand, Place, Rvalue, StatementKind, TerminatorKind,
+};
+use rustc_middle::ty::InternalSubsts;
+use rustc_middle::ty::TyCtxt;
+use rustc_target::abi::VariantIdx;
+
+pub struct ValgrindClientRequest;
+
+impl<'tcx> MirPass<'tcx> for ValgrindClientRequest {
+    #[instrument(skip(self, tcx, body))]
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        if !tcx.sess.opts.unstable_opts.instrument_krabcake {
+            info!("Not instrumenting for krabcake");
+            return;
+        }
+        info!("Instrumenting for krabcake now...");
+        let mut patch = MirPatch::new(body);
+
+        for (block_index, block_data) in body.basic_blocks.iter_enumerated() {
+            for (stmt_index, stmt) in block_data.statements.iter().enumerate() {
+                match &stmt.kind {
+                    StatementKind::Intrinsic(box NonDivergingIntrinsic::Assume(operand)) => {
+                        let loc = Location { block: block_index, statement_index: stmt_index };
+                        info!("Found assume intrinsic (operand={operand:?}. At {loc:?}");
+                        patch = call_ikr(tcx, patch, body, loc, operand);
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        patch.apply(body);
+    }
+}
+
+fn call_ikr<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut patch: MirPatch<'tcx>,
+    body: &Body<'tcx>,
+    loc: Location,
+    _operand: &Operand<'tcx>,
+) -> MirPatch<'tcx> {
+    let span = patch.source_info_for_location(body, loc).span;
+
+    let op = |flag: bool| {
+        Operand::Constant(Box::new(Constant {
+            span,
+            user_ty: None,
+            literal: ConstantKind::from_bool(tcx, flag),
+        }))
+    };
+
+    let (place, rvalue) = {
+        let krabcake_req_did = tcx.lang_items().krabcake_request().unwrap();
+        let krabcake_req_substs = InternalSubsts::identity_for_item(tcx, krabcake_req_did);
+        let krabcake_req_def = tcx.adt_def(krabcake_req_did);
+        let krabcake_req_ty = tcx.mk_adt(krabcake_req_def, krabcake_req_substs);
+        let rvalue = Rvalue::Aggregate(
+            Box::new(AggregateKind::Adt(
+                krabcake_req_did,
+                VariantIdx::new(0),
+                &krabcake_req_substs,
+                None,
+                None,
+            )),
+            vec![op(true)],
```

Comment:
TODO: Get the bool from the existing MIR `intrinsics::assume` itself

Reply:
namely, extract it from the `_operand` parameter you are passing above, right?

Reply:
correct